http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/conn/IdleConnectionReaper.java

```diff
@@ -15,181 +15,163 @@
+    @SdkTestInternalApi
+    IdleConnectionReaper(Map<HttpClientConnectionManager, Long> connectionManagers,
+                         Supplier<ExecutorService> executorServiceSupplier,
+                         long sleepPeriod) {
+
+        this.connectionManagers = connectionManagers;
+        this.executorServiceSupplier = executorServiceSupplier;
+        this.sleepPeriod = sleepPeriod;
     }
 
     /**
-     * Registers the given connection manager with this reaper;
+     * Register the connection manager with this reaper.
      *
-     * @param connectionManager Connection manager to register
-     * @param maxIdleInMs       Max idle connection timeout in milliseconds for this connection manager.
-     * @return true if the connection manager has been successfully registered; false otherwise.
+     * @param manager The connection manager.
+     * @param maxIdleTime The maximum time connections in the connection manager are to remain idle before being reaped.
+     * @return {@code true} If the connection manager was not previously registered with this reaper, {@code false}
+     * otherwise.
      */
-    public static boolean registerConnectionManager(HttpClientConnectionManager connectionManager, long maxIdleInMs) {
-        if (instance == null) {
-            synchronized (IdleConnectionReaper.class) {
-                if (instance == null) {
-                    IdleConnectionReaper newInstance = new IdleConnectionReaper();
-                    newInstance.start();
-                    instance = newInstance;
-                }
-            }
+    public boolean registerConnectionManager(HttpClientConnectionManager manager, long maxIdleTime) {
+        boolean notPreviouslyRegistered = connectionManagers.put(manager, maxIdleTime) == null;
+        if (notPreviouslyRegistered) {
+            setupExecutorIfNecessary();
         }
-        return CONNECTION_MANAGERS.put(connectionManager, maxIdleInMs) == null;
+        return notPreviouslyRegistered;
     }
 
     /**
-     * Removes the given connection manager from this reaper,
-     * and shutting down the reaper if there is zero connection manager left.
+     * Deregister this connection manager with this reaper.
      *
-     * @return true if the connection manager has been successfully removed, false otherwise.
+     * @param manager The connection manager.
+     * @return {@code true} If this connection manager was previously registered with this reaper and it was removed, {@code
+     * false} otherwise.
      */
-    public static boolean removeConnectionManager(HttpClientConnectionManager connectionManager) {
-        boolean wasRemoved = CONNECTION_MANAGERS.remove(connectionManager) != null;
-        if (CONNECTION_MANAGERS.isEmpty()) {
-            shutdown();
+    public boolean deregisterConnectionManager(HttpClientConnectionManager manager) {
+        boolean wasRemoved = connectionManagers.remove(manager) != null;
+        if (wasRemoved) {
```

Comment:
Don't follow this. Shouldn't we only shutdown if all connection managers are de-registered?

Reply:
Woops! Yeah should be an empty check